### PR TITLE
Remove properties from DocumentSpan

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This was always a nasty hack used to pass some information around in Find-References.  Moved to passing the data around explicitly, and making DocumentSpan a simple Document+TextSpan value.

### CODE REVIEW

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptFindUsagesContext.cs [hunk 1]
```diff
@@ -84,8 +86,8 @@ internal VSTypeScriptDefinitionItem(DefinitionItem underlyingObject)
             bool displayIfNoReferences = true)
         {
             return new(DefinitionItem.Create(
-                tags, displayParts, sourceSpans.SelectAsArray(span => span.ToDocumentSpan()), nameDisplayParts,
-                properties: null, displayableProperties: ImmutableDictionary<string, string>.Empty, displayIfNoReferences: displayIfNoReferences));
+                tags, displayParts, sourceSpans.SelectAsArray(span => span.ToDocumentSpan()), sourceSpans.SelectAsArray(_ => (ClassifiedSpansAndHighlightSpan?)null),
+                nameDisplayParts, properties: null, displayableProperties: ImmutableDictionary<string, string>.Empty, displayIfNoReferences: displayIfNoReferences));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptFindUsagesContext.cs:90]: checked TS and F#.  Neither wer using these properties every.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptFindUsagesContext.cs [hunk 2]
```diff
@@ -120,7 +122,8 @@ internal sealed class VSTypeScriptSourceReferenceItem(
         VSTypeScriptDocumentSpan sourceSpan,
         VSTypeScriptSymbolUsageInfo symbolUsageInfo)
     {
-        internal readonly SourceReferenceItem UnderlyingObject = new SourceReferenceItem(definition.UnderlyingObject, sourceSpan.ToDocumentSpan(), symbolUsageInfo.UnderlyingObject);
+        internal readonly SourceReferenceItem UnderlyingObject = new SourceReferenceItem(
+            definition.UnderlyingObject, sourceSpan.ToDocumentSpan(), classifiedSpans: null, symbolUsageInfo.UnderlyingObject);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/IVSTypeScriptFindUsagesContext.cs:126]: ToDocumentSpan was never adding properties to the DocSpan created, so passing 'null' for classifiedSpans is the right thing.

File: src/Features/Core/Portable/FindUsages/AbstractFindUsagesService.ProgressAdapter.cs [hunk 3]
```diff
@@ -37,11 +37,12 @@ public async ValueTask OnReferenceFoundAsync(Document document, TextSpan span, C
...
 
-                var documentSpan = await ClassifiedSpansAndHighlightSpanFactory.GetClassifiedDocumentSpanAsync(
-                    document, span, options.ClassificationOptions, cancellationToken).ConfigureAwait(false);
+                var documentSpan = new DocumentSpan(document, span);
+                var classifiedSpans = await ClassifiedSpansAndHighlightSpanFactory.ClassifyAsync(
+                    documentSpan, classifiedSpans: null, options.ClassificationOptions, cancellationToken).ConfigureAwait(false);
 
                 await _context.OnReferenceFoundAsync(
-                    new SourceReferenceItem(_definition, documentSpan, SymbolUsageInfo.None), cancellationToken).ConfigureAwait(false);
+                    new SourceReferenceItem(_definition, documentSpan, classifiedSpans, SymbolUsageInfo.None), cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FindUsages/AbstractFindUsagesService.ProgressAdapter.cs:45]: both DefinitionItem and SourceReferenceItem can now hold onto these classified spans explicitly. These are then passed along up until FAR looks for them.

File: src/Features/Core/Portable/FindUsages/DefinitionItem.DetachedDefinitionItem.cs [hunk 4]
```diff
@@ -76,18 +77,22 @@ public override int GetHashCode()
...
+            using var convertedClassifiedSpans = TemporaryArray<ClassifiedSpansAndHighlightSpan?>.Empty;
             foreach (var ss in SourceSpans)
             {
                 var documentSpan = await ss.TryRehydrateAsync(solution, cancellationToken).ConfigureAwait(false);
                 if (documentSpan == null)
                     return null;
 
                 converted.Add(documentSpan.Value);
+
+                // todo: consider serializing this data.
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FindUsages/DefinitionItem.DetachedDefinitionItem.cs:89]: serialization never serialized out the properties in a DocumentSpan (how could it, it just saw a dictionary with objects).  So this is no change from before. 

Now that the data is strongly typed and simple structs, we could consdier marshalling.

File: src/Tools/ExternalAccess/FSharp/FSharpDocumentSpan.cs [hunk 5]
```diff
@@ -22,21 +22,12 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.FSharp
...
 
         public FSharpDocumentSpan(Document document, TextSpan sourceSpan)
-            : this(document, sourceSpan, properties: null)
-        {
-        }
-
-        public FSharpDocumentSpan(
-            Document document,
-            TextSpan sourceSpan,
-            ImmutableDictionary<string, object> properties)
```
Comment (CyrusNajmabadi) [inline src/Tools/ExternalAccess/FSharp/FSharpDocumentSpan.cs:35]: i checked and F# never uses this constructor.  so safe to remove.  @vzarytovskii  as a heads up though.

File: src/VisualStudio/Core/Def/FindReferences/Contexts/WithReferencesFindUsagesContext.cs [hunk 6]
```diff
@@ -71,10 +71,13 @@ private async Task AddDeclarationEntriesAsync(DefinitionItem definition, bool ex
 
                 using var _1 = ArrayBuilder<Entry>.GetInstance(out var declarations);
                 using var _2 = PooledHashSet<(string? filePath, TextSpan span)>.GetInstance(out var seenLocations);
-                foreach (var declarationLocation in definition.SourceSpans)
+
+                for (int i = 0, n = definition.SourceSpans.Length; i < n; i++)
                 {
+                    var declarationLocation = definition.SourceSpans[i];
+                    var classifiedSpans = definition.ClassifiedSpans[i];
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/FindReferences/Contexts/WithReferencesFindUsagesContext.cs:78]: FAR ui is the only place we even read out this data.

**Review by genlu — APPROVED** [review]
:shipit:

**Pull request merged**